rstream/client.py

```diff
@@ -131,16 +135,23 @@ def remove_handler(self, frame_cls: Type[FT], name: Optional[str] = None) -> Non
         else:
             self._handlers[frame_cls].clear()
 
+    def get_is_connection_active(self) -> bool:
+        if self._conn is None:
+            return False
+        return True
+
     async def send_frame(self, frame: schema.Frame) -> None:
         logger.debug("Sending frame: %s", frame)
         assert self._conn
         try:
             await self._conn.write_frame(frame)
         except socket.error as e:
+            self._conn = None
             if self._connection_closed_handler is None:
-                print("TCP connection closed")
+                logger.exception("TCP connection closed")
             else:
-                result = self._connection_closed_handler(e)
+                connection_error_info = DisconnectionErrorInfo(e, self._streams)
+                result = self._connection_closed_handler(connection_error_info)
```

Comment:
Actually I tested a Producer scenario like and if I don' trigger there then the call back is no triggered.

I tried this scenario (closing the connection from the server in the UI)

```
async def publish():

    async def on_connection_closed(disconnection_info: DisconnectionErrorInfo) -> None:
        print(
            "connection has been closed from stream: "
            + str(disconnection_info.streams)
            + " for reason: "
            + str(disconnection_info.reason)
        )

        # clean close or reconnect
        global connection_is_closed
        connection_is_closed = True
        await producer.close()



    async with Producer(
        "localhost", username="guest", password="guest", connection_closed_handler=on_connection_closed
    ) as producer:
        # create a stream if it doesn't already exist
        await producer.create_stream(STREAM, exists_ok=True)

        # sending a million of messages in AMQP format
        start_time = time.perf_counter()

        for i in range(MESSAGES):
            amqp_message = AMQPMessage(
                body="hello: {}".format(i),
            )
            # send is asynchronous
            if connection_is_closed is False:
                await producer.send(stream=STREAM, message=amqp_message)
            else:
                break

        end_time = time.perf_counter()
        print(f"Sent {MESSAGES} messages in {end_time - start_time:0.4f} seconds")


asyncio.run(publish())
```

Reply:
I tested the code in #121 with `systemctl stop rabbitmq-server`, and got some exceptions. But with the above modifications, it would be fine.

Reply:
@garvenlee you mean the callback get triggered even if you remove it from line 154? In my case it is not happening if I force the connection to close through the UI

Reply:
I mean the code in #121, it is closed successfully.

Reply:
In your code, `producer.close` is called multiple times. When I removed the `async with`, it’s fine with the above modifications.

Additionally, I got one extra `RuntimeError` when Python garbage collector clears a StreamWriter. I think the reason is that `conn.close` is not called due to `is_not_closed` check in `Client.close`. And after I remove the `is_not_closed` check, and decrease refcount of the StreamWriter in `Connection.close`, then RuntimeError is gone.

```Python
async def close(self):
    …
    self._reader = self._writer = None
```

Reply:
@garvenlee Not sure but I'm having issues with my tests removing this call here.

You can see tests in this PR (I also updated the branch with the last suggestions). they are working fine but If I remove the call here the on_connection_closed get not triggered (you can test the producer scenario).

Reply:
Here is my modification, I tested your code above and the code in #121, all is fine.

```Python
# Connection
async def close(self):
    …
    # add this line
    self._reader = self._writer = None


# BaseClient
async def _listener(self):
    while True:
        try:
            frame = await self._conn.read_frame()
        except (ConnectionClosed, socket.error) as exc:
            self._is_not_closed = False
            if self._connection_closed_handler is not None:
                err_info = DisconnectionErrorInfo(exc, self._streams)
                maybe_coro = self._connection_closed_handler(err_info)
                if maybe_coro is not None and inspect.isawaitable(maybe_coro):
                    asyncio.create_task(maybe_coro)
            else:
                logger.exception(“TCP connection closed.”)
            break
        except asyncio.CancelledError:
            logger.info(“close is called.”)
            break
        except BaseException as exc:  # frame error? KeyboardInterrupt or sth.
            logger.exception(f“unknown error: {exc}”)
            break
        else:
            # keep the same
             ……


# BaseClient
async def close(self):
    if self._is_not_closed and self.is_started:
        try:
            await self.sync_request(
                schema.Close(…), 
                resp_schema=schema.CloseResponse, 
                timeout=5,
            )
        except asyncio.TimeoutError:
            print(“Close in sync_request: Timeout”)
        except BaseException as exc:  # Connection is lost
            print(f“Close in sync_request: {exc}”)

    tasks = []
    for task in self._tasks.values():
        task.cancel()
        tasks.append(task)
    await asyncio.wait(tasks)

    await self._conn.close()
    self._conn = None
    self._tasks.clear()
    self._handlers.clear()
    self.server_properties = None
    self._is_not_closed = True

# BaseClient
# `stop_task` isn’t used.
async def _heartbeat_sender(self):
    if self._heartbeat == 0:
        return
    while self._is_not_closed:
        await self.send_frame(…)
        await asyncio.sleep(delay)
        if - statement
        # no suppress here, it will be cancelled in `close`, and `asyncio.wait` controls this.

# Client
# remove `close` function
```

About Producer:

```Python
async def close(self):
    if self._default_client is None:
        return

    self._close_called = True
    if self.task is not None:
        task, self.task = self.task, None
        await task  # wait `_timer` done, better to add double check in it

    for publisher in self._publishers.values():
        client = publisher.client        
        if client.is_connection_alive():
            try:
                client.delete_publisher(publisher.id)
            except asyncio.TimeoutError:
                print(…)
            except BaseException as exc:
                print(…)
        client.remove_handler()
        client.remove_handler()
    self._publisher.clear()
    await self._pool.close()
    self._clients.clear()
    self._waiting_for_confirm.clear()
    self.default_client = None
    self._close_called = False
```